File: Seori/Weekly Contest 409/3243. Shortest Distance After Road Addition Queries I.py
class Solution:
    def shortestDistanceAfterQueries(self, n: int, queries: List[List[int]]) -> List[int]:
        
        # [A] 0부터 n-1까지의 도시를 연결하는 최단 거리를 찾는 bfs 함수를 정의한다.
        #     Define a bfs function to find the shortest distance connecting cities from 0 to n-1.
        def find(start: int) -> int:
            visited = [0] * (n-1)
            queue = deque([(start, 0)])
            visited[start] = 1
            while queue:
                now, distance = queue.popleft()
                for next in cities[now]:

                    # [A-1] 목적지에 도달했다면 거리를 반환한다.
                    #       If you have reached your destination, return the distance.
                    if next == n - 1:
                        return distance + 1
                    
                    # [A-2] 목적지에 도달할 때까지 다음 도시로 이동한다.
                    #       Move to the next city until you reach your destination.
                    if not visited[next]:
                        queue.append((next, distance + 1))
                        visited[next] = 1

        # [1] 2차원 배열 cities에 각 도시의 연결 정보를 저장한다.
        #     Save the connection information of each city in the 2D array cities.
        cities = []
        for i in range(n-1):
            cities.append([i+1])

        # [2] 각 쿼리마다 도시의 연결 정보를 업데이트하고 최단 거리를 찾는다.
        result = []
        for i, j in queries:
            cities[i].append(j)
            result.append(find(0))

        return result

File: Gombang/Weekly Contest 409/3243_Shortest_Distance_After_Road_Addition_Queries_I.cs
using System.Collections.Generic;

public class Solution
{
	public int[] ShortestDistanceAfterQueries(int n, int[][] queries)
	{
		int[] result = new int[queries.Length];

		// queries에 대한 다리를 제외하고 n까지 이루어진 노드를 연결.
		List<List<int>> bridge = new List<List<int>>();
		for (int i = 0; i < n - 1; i++)
		{
			bridge.Add(new List<int>());
			bridge[i].Add(i + 1);
		}

		for (int i = 0; i < queries.Length; i++)
		{
			int newBridgeStartIndex = queries[i][0];
			int newBridgeEndIndex = queries[i][1];

			bridge[newBridgeStartIndex].Add(newBridgeEndIndex);

			// 새로 연결한 다리를 기준으로 BFS를 진행.
			result[i] = BFS(bridge, n);
		}

		return result;
	}

	private int BFS(List<List<int>> graph, int n)
	{
		Queue<(int node, int distance)> queue = new Queue<(int node, int distance)>();
		bool[] visited = new bool[n];

		// 0번 도시에서 시작
		queue.Enqueue((0, 0));
		visited[0] = true;

		while (queue.Count > 0)
		{
			var (node, distance) = queue.Dequeue();

			// n-1번 도시에 도착한 경우
			if (node == n - 1)
				return distance;

			// 인접 노드 탐색
			foreach (int neighbor in graph[node])
			{
				if (!visited[neighbor])
				{
					visited[neighbor] = true;
					queue.Enqueue((neighbor, distance + 1));
				}
			}
		}

		// n-1에 도달할 수 없는 경우
		return -1;
	}
}